SpatialGDK/Source/Private/Interop/SpatialView.cpp

```diff
@@ -125,50 +125,63 @@ Worker_Authority USpatialView::GetAuthority(Worker_EntityId EntityId, Worker_Com
 	return WORKER_AUTHORITY_NOT_AUTHORITATIVE;
 }
 
-improbable::UnrealMetadata* USpatialView::GetUnrealMetadata(Worker_EntityId EntityId)
+template <typename T>
+typename T* USpatialView::GetComponentData(Worker_EntityId EntityId)
 {
-	if (TSharedPtr<improbable::UnrealMetadata>* UnrealMetadataPtr = EntityUnrealMetadataMap.Find(EntityId))
+	if (TMap<Worker_ComponentId, TSharedPtr<ComponentStorageBase>>* ComponentStorageMap = EntityComponentMap.Find(EntityId))
 	{
-		return UnrealMetadataPtr->Get();
-	}
-
-	return nullptr;
-}
-
-improbable::EntityAcl* USpatialView::GetEntityACL(Worker_EntityId EntityId)
-{
-	if (TSharedPtr<improbable::EntityAcl>* EntityAclPtr = EntityACLMap.Find(EntityId))
-	{
-		return EntityAclPtr->Get();
+		if (TSharedPtr<improbable::ComponentStorageBase>* Component = ComponentStorageMap->Find(T::ComponentId))
+		{
+			return &static_cast<improbable::ComponentStorage<T>&>(*(Component->Get())).Get();
+		}
 	}
 
 	return nullptr;
 }
 
 void USpatialView::OnAddComponent(const Worker_AddComponentOp& Op)
 {
-	if (Op.data.component_id == improbable::UnrealMetadata::ComponentId)
+	TSharedPtr<improbable::ComponentStorageBase> Data;
+	switch (Op.data.component_id)
 	{
-		EntityUnrealMetadataMap.Add(Op.entity_id, MakeShared<improbable::UnrealMetadata>(Op.data));
+	case SpatialConstants::ENTITY_ACL_COMPONENT_ID:
+		Data = MakeShared<improbable::ComponentStorage<improbable::EntityAcl>>(Op.data);
+		break;
+	case SpatialConstants::METADATA_COMPONENT_ID:
+		Data = MakeShared<improbable::ComponentStorage<improbable::Metadata>>(Op.data);
+		break;
+	case SpatialConstants::POSITION_COMPONENT_ID:
+		Data = MakeShared<improbable::ComponentStorage<improbable::Position>>(Op.data);
+		break;
+	case SpatialConstants::PERSISTENCE_COMPONENT_ID:
+		Data = MakeShared<improbable::ComponentStorage<improbable::Persistence>>(Op.data);
+		break;
+	case SpatialConstants::ROTATION_COMPONENT_ID:
+		Data = MakeShared<improbable::ComponentStorage<improbable::Rotation>>(Op.data);
+		break;
+	case SpatialConstants::UNREAL_METADATA_COMPONENT_ID:
+		Data = MakeShared<improbable::ComponentStorage<improbable::UnrealMetadata>>(Op.data);
+		break;
+	default:
+		return;
 	}
 
-	if (Op.data.component_id == improbable::EntityAcl::ComponentId)
-	{
-		EntityACLMap.Add(Op.entity_id, MakeShared<improbable::EntityAcl>(Op.data));
-	}
+	EntityComponentMap.FindOrAdd(Op.entity_id).FindOrAdd(Op.data.component_id) = Data;
 }
 
 void USpatialView::OnRemoveEntity(const Worker_RemoveEntityOp& Op)
 {
-	EntityUnrealMetadataMap.Remove(Op.entity_id);
-	EntityACLMap.Remove(Op.entity_id);
+	EntityComponentMap.Remove(Op.entity_id);
 }
 
 void USpatialView::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
 {
-	if (Op.update.component_id == improbable::EntityAcl::ComponentId)
+	if (Op.update.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID)
```

Comment:
Why this change?

Reply:
Just for consistency with `OnAddComponent` which also uses `SpatialConstants`.